src/mpl/include/mpl_gpu.h

```diff
@@ -51,6 +51,13 @@ static inline int MPL_gpu_query_pointer_attr(const void *ptr, MPL_pointer_attr_t
 
 int MPL_gpu_query_support(MPL_gpu_type_t * type);
 int MPL_gpu_query_pointer_attr(const void *ptr, MPL_pointer_attr_t * attr);
+static inline bool MPL_gpu_query_pointer_is_dev(const void *ptr)
```

Comment:
Added this helper in MPL in this new version of the PR. A potential issue is that this query will happen twice in the communication path if the buffer lives on the GPU. Once in the netmod, and again in the generic layer to choose the right code path.

Nevertheless, this fixes a potential crash (bug) in GPU-aware builds, as reported in #4832.

Reply:
It's okay for now. But we need to think about how to avoid the multiple queries in a single communication path. One way is to modify the ADI interface so we use a struct rather than the current `buf, count, datatype`. The struct can store the extra info such as `MPL_pointer_attr_t`. An alternative way is to implement some cache system. If the goal is just to avoid multiple queries in a single chain of call path, we can implement the cache on per-thread storage and always clear out the cache at the end of call path, so the cost of cache can be made minimal. If a simple per-thread stack can work, then the cache system can be really simple and efficient.

Reply:
Can we add the helper in MPIR rather than MPL? See `MPIR_GPU_query_pointer_attr`. It allows us to disable the query even with --with-cuda build by setting MPIR_CVAR_ENABLE_GPU=0.